File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/popup/PopupWidget.java
package com.sup.dev.android.views.popup;

import com.sup.dev.android.views.widgets.Widget;
import com.sup.dev.android.views.widgets.WidgetViewWrapper;

public class PopupWidget extends Popup implements WidgetViewWrapper {

    private final Widget widget;

    public PopupWidget(Widget widget){
        super(widget.getView());
        this.widget = widget;
        setCancelable(widget.isCancelable());
        setEnabled(widget.isEnabled());
    }

    @Override
    protected void onShow() {
        super.onShow();
        widget.onShow();
    }

    @Override
    protected void onHide() {
        super.onHide();
        widget.onHide();
    }

    @Override
    public PopupWidget hide() {
        return super.hide();
    }

    //
    //  Setters
    //


    public PopupWidget setCancelable(boolean cancelable) {
        return super.setCancelable(cancelable);
    }

    public PopupWidget setEnabled(boolean enabled) {
        return super.setEnabled(enabled);
    }

    @Override
    public PopupWidget hideWidget() {
        return hide();
    }

    @Override
    public PopupWidget setWidgetCancelable(boolean cancelable) {
        return setCancelable(cancelable);
    }

    @Override
    public PopupWidget setWidgetEnabled(boolean enabled) {
        return setEnabled(enabled);
    }
}

File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/popup/PopupWidget.kt
package com.sup.dev.android.views.popup

import com.sup.dev.android.views.widgets.Widget
import com.sup.dev.android.views.widgets.WidgetViewWrapper


class PopupWidget(private val widget: Widget) : Popup(widget.view), WidgetViewWrapper {

    init {
        setCancelable<Popup>(widget.isCancelable)
        setEnabled<Popup>(widget.isEnabled)
    }

    override fun onShow() {
        super.onShow()
        widget.onShow()
    }

    override fun onHide() {
        super.onHide()
        widget.onHide()
    }

    override fun <K : Popup> hide(): K {
        dismiss()
        return this as K
    }

    //
    //  Setters
    //

    override fun <K : Popup> setCancelable(cancelable: Boolean): K {
        super.setCancelable<Popup>(cancelable)
        return this as K
    }

    override fun <K : Popup> setEnabled(enabled: Boolean): K {
        super.setEnabled<Popup>(enabled)
        return this as K
    }

    override fun <K : WidgetViewWrapper> hideWidget(): K {
        return hide<Popup>() as K
    }

    override fun <K : WidgetViewWrapper> setWidgetCancelable(cancelable: Boolean): K {
        return setCancelable<Popup>(cancelable) as K
    }

    override fun <K : WidgetViewWrapper> setWidgetEnabled(enabled: Boolean): K {
        return setEnabled<Popup>(enabled) as K
    }


}
